### ISSUE TITLE
Code generation failing when having two associations to the same class. 

### ISSUE THREAD

**Issue by Nava2** [umple/umple#188]
Originally reported on Google Code with ID 288

---

 namespace Assignment2;

 class Person{
 }

 class Course{
- -> 1 Person prof;
- course -- 1..\* Person student;
  }

Here we have two associations from the class Course to class Person. One is many-to-one
and the other many-to-many.
The Java code generated contains compilation errors in many places, particularly in
the setStudent(...) method.

---

Reported by @mgarzon on 2012-03-25 21:10:44

**Comment by Nava2** [comment]
I am increasing the priority of this and also giving it additional keywords

---

Reported by @umple on 2012-03-27 22:07:56
- **Labels added**: Priority-High, Component-Metamodel, associations
- **Labels removed**: Priority-Medium

**Comment by Nava2** [comment]
Attached generated code, abstracted into two separate files:  Course.java and Person.java.

Below are the errors when an attempt to compile the java code is made:

# 

Course.java:167: newStudent is already defined in setStudent(Person...)
    for (Person newStudent : verifiedStudent)
                     ^
Course.java:180: oldStudent is already defined in setStudent(Person...)
    for (Person oldStudent : oldStudent)
                     ^
Course.java:180: foreach not applicable to expression type
    for (Person oldStudent : oldStudent)
                                         ^
Person.java:68: cannot find symbol
symbol  : variable courses
location: class Person
    if (courses.contains(aCourse)) { return false; }
        ^
4 errors

# 

The last error is blatantly obvious, the class attribute "course" is being accessed
by the name "courses" which is an error.  The fix is "course" instead of "courses"
here.

The others seem to be errors and inconsistencies in naming.  A few are not.  I'm wondering
if there are other errors for different associations.

Investigating.

---

Reported by @EmperorJordan on 2012-03-28 15:48:33
- _Attachment: [Course.java](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-2/Course.java)_
- _Attachment: [Person.java](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-2/Person.java)_

**Comment by Nava2** [comment]
Manually fixed the compilation errors simply by changing the names of the variables...
naming conflicts it seems but I am wary on this.

Clearly the problem is in the jet templates and that is where the investigation should
focus.

---

Reported by @EmperorJordan on 2012-03-28 15:59:37

**Comment by Nava2** [comment]
I have  determined that this problem is not related to two associations. It is actually
related to a single 1--\* association with a role name. The following cases both give
rise to it:

There is clearly a longstanding bug in one of the template cases. I am a bit shocked
that we never had a test for this combination. We obviously need more test cases. I
have added both Jordan and Sonya to the cc list since I suggested that either/both
of them could tackle it if they had time.

namespace Assignment2;

 class Person {
 }

 class Course{
  1..\* cour -- \* Person;
 }

---

namespace Assignment2;

 class Person {
 }

 class Course{
- cour -- 1..\* Person human;
  }

---

Reported by @umple on 2012-03-28 18:54:03
- **cc added:** +@EmperorJordan, +`perpetualcyclist`

**Comment by Nava2** [comment]
I am almost positive this is a naming conflict only problem.  Taking your first example,
I was able to solve the issues:

Person.java:147: newCour is already defined in setCour(Course...)
    for (Course newCour : verifiedCour)
                ^
Person.java:160: oldCour is already defined in setCour(Course...)
    for (Course oldCour : oldCour)
                ^
Person.java:160: foreach not applicable to expression type
    for (Course oldCour : oldCour)
                          ^

Which were name changes.  I've attached the modified generated code for the first case
(Person.java and Course.java).  Comparing this with the generated code there is little
difference, as everything was in the "setSour()" method and the changes were only to
resolve the naming conflicts.

Are the changes valid, or does this break intended behaviour?  Clearly it will compile
with the new generated code I modified, but correct behaviour has to be ensured.

Thanks

---

Reported by @EmperorJordan on 2012-04-01 04:03:57
- _Attachment: [Course.java](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-5/Course.java)_
- _Attachment: [Person.java](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-5/Person.java)_

**Comment by Nava2** [comment]
I've done the same thing as with the first example to the second example - applied the
naming conflict fix and it seems to compile now.  The errors were very similar:

Course.java:147: newHuman is already defined in setHuman(Person...)
    for (Person newHuman : verifiedHuman)
                ^
Course.java:160: oldHuman is already defined in setHuman(Person...)
    for (Person oldHuman : oldHuman)
                ^
Course.java:160: foreach not applicable to expression type
    for (Person oldHuman : oldHuman)
                           ^

I've attached the modified generated code files for verification if the fix breaks
intended behaviour.  I think the next move, assuming this is correct, is to make the
change in the appropriate jet template(s).

---

Reported by @EmperorJordan on 2012-04-01 04:09:41
- _Attachment: [Person.java](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-6/Person.java)_
- _Attachment: [Course.java](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-6/Course.java)_

**Comment by Nava2** [comment]
I've looked at this issue, and in my opinion, all the errors are caused very simply
by the problem of _pluralization_, about which the jet templates make some assumptions:

1) The association `* course -- 1..* Person student` has a _singular_ name, even though
it technically represents a "to many" relationship. None of our test cases test this
situation. Internally, inside the `setStudent` method, the singular `newStudent` variable
name is used in the loop, with the assumption that the method parameter would be named
`newStudent*s*` since this method is the setter for a "to many" relationship. Clearly
that assumption is incorrect in this case since the association still has a singular
name. I think that instead of using the "new___" singular variable name internally
-- which is sometimes also used by the Jet templates as the name for a method parameter
-- an additional variable name should be added to the generator's dictionary that is
ONLY used internally within methods, e.g. `tmpStudent`, `tmpStudents`, `tmpNewStudent`,
`tmpOldStudent`, etc... That, I suspect, is the best and most conceptually correct
way to prevent naming collisions of the sort that are causing the errors on lines 147
and 160 of Course.java. There are a number of places in the jet templates where such
a change would have to be made.

2) The error on line 68 of Person.java is also a pluralization problem. Again, it's
a "to many" association end that is given a singular association name. If I rename
the association end to "courses", all errors are fixed. The pluralization problem is
basically handled correctly by this method, except that for some reason the logic of
the initial check (to see if the collection already contains the item) is duplicated,
and one instance of the duplicated logic handles the pluralization correctly whereas
the other doesn't. One will need to find the location in the jet templates where this
duplication is happening, determine what other test cases the duplicated code might
be affecting, and find the best way to fix the problem.

---

Reported by `perpetualcyclist` on 2012-04-11 04:45:19

**Comment by Nava2** [comment]
Clarification on point (2) in last comment: if it wasn't clear, what I meant to say
is that the variables seem to be named correctly within the method -- regardless of
whether the association name is plural or not -- except in the one line of duplicated
logic.

---

Reported by `perpetualcyclist` on 2012-04-11 05:54:01

**Comment by Nava2** [comment]
All variable names (should) flow through helper methods, so there shouldn't
be many changes to the JET templates.  Within those helper methods it
should properly present the appropriate singular versus pluralization of
the word.

Anywhere that is not the case, I would update the call to use that helper
method.

The change to "tmp" is a separate change (one which I am not that fond of)
- I would first focus the plural issue and then look at the naming
  convention.

---

Reported by @aforward on 2012-04-11 11:53:13

**Comment by Nava2** [comment]
Reported by @umple on 2012-06-14 15:00:24
- **Labels added**: Diffic-Easy

**Comment by Nava2** [comment]
This is a long bug.  Here is what I think should be the correct line of action.

# 

# ISSUE

Umple does not detect duplicate instance variable names.  Just like in Java you cannot
name an instance variable the same, in Umple a similar error should be raised.

# 

# RESOLUTION

Umple parser should check classes for duplicate names (either duplicate attributes,
or association ends) and raise an error if a duplicate is encountered.

The first example above shows "one" way to do this for associations as both associations
are named the name (the default name for "Course" is course, so the \* -- end without
a name is duplicately named with the subsequent \* course association)

# 

# TEST CASES

Both examples should raise errors:

{{{
class Person {

  Integer one;
  Integer one;

 }
}}}

And, the original

{{{
class Person{
 }

 class Course{
- -> 1 Person prof;
- course -- 1..\* Person student;
  }

}}}

---

Reported by @aforward on 2012-06-14 15:34:51

**Comment by Nava2** [comment]
Reported by @umple on 2012-09-27 19:40:09
- **Owner removed**

**Comment by Nava2** [comment]
See also issue 172 - another 'duplicate element' issue

---

Reported by @umple on 2012-09-27 20:30:00

**Comment by Nava2** [comment]
Some aspects of this issue have now been dealt with. However the core issue remains.

The following is a test case that generates multiple methods with the same signature:
class Person{
 }

 class Course{
- -- 1 Person prof;
- course -- 1..\* Person student;
  }

---

Reported by @umple on 2012-12-31 13:58:26
- **Labels added**: ucosp
- **cc removed:** -@aforward, -@umple, -@EmperorJordan, -`perpetualcyclist`

**Comment by Nava2** [comment]
Reported by @umple on 2013-01-30 13:54:31
- **Owner changed**: `robinjay87`

**Comment by Nava2** [comment]
Spun off issue 178 as a separate issue from this.

The core test case now becomes:

// Generates multiple methods with the same signature

class Person{
 }

class Course{
- -- 1 Person prof;
- course -- 1..\* Person students;
  }

When two associations have the same role name, detect this better in
the parser and require one role name to be changed. I think this would be
the simplest solution. You would just improve the code that issues error
message 19. The above case was just omitted.

here is another:

class A {
  1 first -- \* B b;
  1 second --\* B ;
}

class B {
}

In general, it seems that in the code for message 19 detects if there is
no role name in the above, or if one of the role names is pluralzed 'bs', 
but not if one is singular 'b', and it turns out that methods generated  
from the names clash in either case.

---

Reported by @umple on 2013-02-19 19:27:00

**Comment by Nava2** [comment]
Here is another strange aspect of this issue:

class Person{}

class Course{
  // Association A
  // \* -> 1 Person prof;

  // Association B
- course -- 1..\* Person student;
  }

If I comment out association B all is OK.

If I comment out association A I get issue 277 that I raised earlier.

If I leave both associations in, then I get issue 277 plus the extra 
problem you reported

If I change 
- course
  to
- courses then the extra line is generated (for no reason) but does not  
  give a java error (issue 277 is still there).

If I change
- course
  to
- whatever

Then the extra line is not generated.

---

Reported by @umple on 2013-02-20 02:39:04

**Comment by Nava2** [comment]
To clarify the last comment: The two associations seem to be interacting; and are dependent
on the presence of a role name 'course' or 'courses' in association B, which causes
 addCourse to have the following lines if the role name is called 'course'

```
if (courses.contains(aCourse)) { return false; }
if (course.contains(aCourse)) { return false; }
```

(the first of which is invalid because the variable is called 'course' due to the role
name).

It would seem that the complex logic for selecting the method to output is failing,
but it is a deep mystery why it matters that there there be two associations to trigger
this.

---

Reported by @umple on 2013-02-20 02:44:22

**Comment by Nava2** [comment]
These test cases have been covered with the patch I have attached:
class Person{
 }

class Course{
- -- 1 Person prof;
- course -- 1..\* Person students;
  }
  ------------------------------
  class A {
  1 first -- \* B b;
  1 second --\* B ;
  }

class B {

## }

HOWEVER, there still remains the test case, which creates unique errors: 

class Person{}

class Course{
  // Association A
  // \* -> 1 Person prof;

  // Association B
- course -- 1..\* Person student;
  }

(more details - see Feb 19, 2013 entry from Tim)

---

Reported by `robinjay87` on 2013-03-12 17:33:07
- **Status changed**: `MostlyDone`
  - _Attachment: [patch-RobinJ-288-roleNameSameAsClassWithMultiAssocToSameClass](https://storage.googleapis.com/google-code-attachments/umple/issue-288/comment-19/patch-RobinJ-288-roleNameSameAsClassWithMultiAssocToSameClass)_

**Comment by Nava2** [comment]
Reported by `redeyedmars` on 2013-05-06 15:26:04
- **Owner removed**
- **Labels removed**: ucosp

**Comment by Nava2** [comment]
Reported by @umple on 2013-09-12 15:06:06
- **Labels added**: ucosp

**Comment by Nava2** [comment]
I cannot reproduce this any more,. so it seems fixed.

---

Reported by @umple on 2013-09-17 14:35:04
- **Status changed**: `Verified`
